### PR TITLE
Some mixed recipes

### PR DESCRIPTION
**Pull request by robertodr**
Some recipes from chapter 7 and some recipes from chapter 10. Not typecloud-ed, but one has to start somewhere! 
Specifically:
1. Chapter 7, recipe 1. Where I discuss some basics about the `install` command.
2. Chapter 7, recipe 2. Which is the same as recipe 1, but where I show usage of `GNUInstallDirs`.
3. Chapter 7, recipe 3. Where I discuss how to generate the export header and hiding of symbols. This recipe probably belongs somewhere else...
4. Chapter 10, recipe 3. Where I show how to use `ExternalProject` to download, configure and build a specific version of Boost at build time.
5. Chapter 10, recipe 4. Where a superbuild is demonstrated by means of FFTW3.

### CODE REVIEW

**Comment by bast** [comment]
Thanks! I will review today.

File: Chapter07/recipe-03/cxx-example/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  file(WRITE "${Boost_USER_CONFIG_JAMFILE}" "using ${toolset} : : ${CMAKE_CXX_COMPILER} : <cxxflags>\"-std=c++11\" <linkflags>\"-std=c++11\" ;")
+  set(Boost_BUILD_USER_CONFIG_ARGS "--user-config=${PROJECT_BINARY_DIR}/user-config.jam")
+
+  ExternalProject_Add(custom_boost
+    PREFIX
+      external
+    URL
+      https://dl.bintray.com/boostorg/release/1.64.0/source/boost_1_64_0.zip
+    URL_HASH
+      SHA256=b99973c805f38b549dbeaf88701c0abeff8b0e8eaa4066df47cac10a32097523
```
Comment (bast) [inline Chapter07/recipe-03/cxx-example/CMakeLists.txt]: Nice! I need to remember to use this.

File: Chapter07/recipe-03/cxx-example/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+     set(Boost_${_COMP}_LIBRARY_RELEASE ${Boost_LIBRARY_DIR}/${Boost_${_COMP}_LIBRARY} CACHE FILEPATH "Boost ${_component} library (release)" FORCE)
+     list(APPEND Boost_LIBRARIES ${Boost_${_COMP}_LIBRARY})
+   endforeach()
+
+   set(Boost_INCLUDE_DIRS ${Boost_INCLUDE_DIR})
+   set(Boost_LIBRARY_DIRS ${Boost_LIBRARY_DIR})
+endif()
+
+add_executable(path-info path-info.cpp)
+add_dependencies(path-info custom_boost)
```
Comment (bast) [inline Chapter07/recipe-03/cxx-example/CMakeLists.txt]: I think this might be the preferred mechanism: https://cmake.org/cmake/help/v3.2/module/ExternalProject.html#command:externalproject_add_stepdependencies

File: Chapter10/recipe-02/cxx-example/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
+
+project(recipe-01 CXX)
```
Comment (bast) [inline Chapter10/recipe-02/cxx-example/CMakeLists.txt]: copy-paste

File: Chapter07/recipe-04/c-example/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+set(CMAKE_C_STANDARD 99)
+set(CMAKE_C_EXTENSIONS OFF)
+set(CMAKE_C_STANDARD_REQUIRED ON)
+
+include(GNUInstallDirs)
+include(ExternalProject)
+
+if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+  set(CMAKE_INSTALL_PREFIX "/usr/local/${PROJECT_NAME}" CACHE PATH "Install path" FORCE)
```
Comment (bast) [inline Chapter07/recipe-04/c-example/CMakeLists.txt]: why is this needed? ignore my question if the book will give more context.

File: Chapter10/recipe-01/cxx-example/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
+
+project(recipe-01 CXX)
+
+if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+  set(CMAKE_INSTALL_PREFIX "/usr/local/message" CACHE PATH "Install path" FORCE)
```
Comment (bast) [inline Chapter10/recipe-01/cxx-example/CMakeLists.txt]: Also here looks hard-coded to me but perhaps the book will explain it.

File: Chapter10/recipe-03/cxx-example/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,17 @@
+cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
+
+project(recipe-01 CXX)
+
+if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+  set(CMAKE_INSTALL_PREFIX "/usr/local/message" CACHE PATH "Install path" FORCE)
```
Comment (bast) [inline Chapter10/recipe-03/cxx-example/CMakeLists.txt]: Same as above.

File: Chapter07/recipe-03/cxx-example/CMakeLists.txt (hunk 2 above)
Comment (bast) [reply Chapter07/recipe-03/cxx-example/CMakeLists.txt]: Sorry, correction to my suggestion. It is preferred if we need more fine-grain. I read the docs too quickly. Probably better to leave it as is.

File: Chapter10/recipe-02/cxx-example/CMakeLists.txt (hunk 3 above)
Comment (robertodr) [reply Chapter10/recipe-02/cxx-example/CMakeLists.txt]: The original sin of programming 😸

File: Chapter07/recipe-04/c-example/CMakeLists.txt (hunk 4 above)
Comment (robertodr) [reply Chapter07/recipe-04/c-example/CMakeLists.txt]: To make sure it's set to something sensible. This is suggested in the CMake docs.

Comment (bast) [reply Chapter07/recipe-04/c-example/CMakeLists.txt]: OK thanks! We need to figure out how these things behave on Windows at some point :-)

Comment (robertodr) [reply Chapter07/recipe-04/c-example/CMakeLists.txt]: 😱

**Comment by bast** [comment]
What do we do with the failing tests?

**Comment by robertodr** [comment]
Apologies for not being more attentive to review comments. I'll fix the copy-paste later and also give a stab at CI. Seems a mix of missing packages and wrong compiler flags.

**Comment by bast** [comment]
That's fine. The other developments are independent. It's just that I think it's better if we restore tests before going to master because later it takes lot of energy to get things green again and the "known" red may mask the unknown red.

**Comment by robertodr** [comment]
I am on it.

**Comment by robertodr** [comment]
Remaining failures on Travis are due to Mac OS X (surprise, surprise) The test of the Boost.Python example fails...

**Comment by robertodr** [comment]
@bast let's get these in. I'll patch them up if anything comes up while writing.
